pkg/models/protocoldiscovery.go

```diff
@@ -1,20 +1,28 @@
 // -*- Mode: Go; indent-tabs-mode: t -*-
 //
 // Copyright (C) 2018 Canonical Ltd
+// Copyright (C) 2020 IOTech Ltd
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package models
 
+import contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+
 // ProtocolDiscovery is a low-level device-specific interface implemented
 // by device services that support dynamic device discovery.
 type ProtocolDiscovery interface {
 	// Discover triggers protocol specific device discovery, which is
-	// a synchronous operation which returns a list of new devices
-	// which may be added to the device service based on service
-	// config. This function may also optionally trigger sensor
-	// discovery, which could result in dynamic device profile creation.
-	//
-	// TODO: add models.ScanList (or define locally) for devices
-	Discover() (devices *interface{}, err error)
+	// an asynchronous operation, returns a list of new devices
+	// that may be added to the device service based on a set of acceptance
+	// criteria (i.e. Provision Watchers).
+	Discover(devices chan<- []DiscoveredDevice)
```

Comment:
The previous signature of this method returned an interface pointer, and an error. Just wondering if it makes sense to keep the error as a retval? Otherwise if somehow discovery can't be triggered, we don't still spawn a new goroutine for `filterAndAdd`.

Reply:
Because Discovery will run in another go routine(so that REST call may return immediately), it's hard to wait for its return value.  
We should update the README to notify the user to handle/log the error themself and must push an empty slice of `DiscoveredDevice` so that `filterAndAdd` won't be blocked forever.  
Is this make sense to you ?

Reply:
The SDK could call the implementation synchronously. It would then be up to the implementation whether to start a new goroutine if this would be necessary in order to return in reasonable time. Alternatively, always run in a new goroutine and don't allow for the implementation to return an error (this is effectively what the CSDK does)